kahuna/public/js/components/gr-collections-panel/gr-collections-panel.js

```diff
@@ -30,22 +35,136 @@ grCollectionsPanel.controller('GrCollectionsPanelCtrl', [
         ctrl.error = true;
     }).catch(() => ctrl.error = true);
 
+    ctrl.selectedImages$ = selectedImages$;
 }]);
 
-grCollectionsPanel.controller('GrNodeCtrl', ['collections', function(collections) {
+grCollectionsPanel.factory('collectionsTreeState', ['$window', function($window) {
+    const localStorageKey = 'collectionsTreeState';
+    const defaultShowState = true;
+
+    function getStateFromStorage() {
+        const jsonStr = $window.localStorage.getItem(localStorageKey) || '{}';
+        return JSON.parse(jsonStr);
+    }
+
+
+    function setState(pathId, show) {
+        const newState = angular.extend({}, getStateFromStorage(), {[pathId]: show});
+        $window.localStorage.setItem(localStorageKey, JSON.stringify(newState));
+    }
+
+    function getState(pathId) {
+        const state = getStateFromStorage();
+        return angular.isDefined(state[pathId]) ? state[pathId] : defaultShowState;
+    }
+
+    return {
+        setState,
+        getState
+    };
+
+}]);
+
+grCollectionsPanel.controller('GrNodeCtrl',
+    ['$scope', 'collections', 'subscribe$', 'inject$', 'onValChange', 'collectionsTreeState',
+    function($scope, collections, subscribe$, inject$, onValChange, collectionsTreeState) {
 
     const ctrl = this;
+    const pathId = ctrl.node.data.content.pathId;
+
+    ctrl.saving = false;
+    ctrl.editing = false;
     ctrl.deletable = false;
+    ctrl.showChildren = collectionsTreeState.getState(pathId);
     ctrl.addChild = childName => collections.addChildTo(ctrl.node, childName);
     collections.isDeletable(ctrl.node).then(d => ctrl.deletable = d);
 
     ctrl.remove = () => collections.removeFromList(ctrl.node, ctrl.nodeList);
 
+    ctrl.getCollectionQuery = path => getCollection(path);
+
+    // TODO: move this somewhere sensible, we probably don't want an observable for each node.
+    const add$ = new Rx.Subject();
+    const pathWithImages$ =
+            add$.withLatestFrom(ctrl.selectedImages$, (path, images) => ({path, images}));
+
+    const hasImagesSelected$ = ctrl.selectedImages$.map(i => i.size > 0);
+    ctrl.addImagesToCollection = () => {
+        ctrl.saving = true;
```

Comment:
Could this live in the `subscribe$` too to isolate side-effects and group the changes to `ctrl.saving` in the same code location?